optimade/filtertransformers/mongo.py

```diff
@@ -386,30 +387,55 @@ def replace_with_relationship(subdict, prop, expr):
                     f'Cannot filter relationships by field "{_field}", only "id" is supported.'
                 )
 
-            # in the case of HAS ONLY, the size operator needs to be applied
-            # one level up, i.e. excluding the field
-            if "$size" in expr:
-                if "$and" not in subdict:
-                    subdict["$and"] = []
-                subdict["$and"].extend(
-                    [
-                        {
-                            f"relationships.{_prop}.data": {
-                                "$size": expr.pop("$size"),
-                            }
-                        },
-                        {f"relationships.{_prop}.data.{_field}": expr},
-                    ]
+            subdict[f"relationships.{_prop}.data.{_field}"] = expr
+            subdict.pop(prop)
+            return subdict
+
+        return recursive_postprocessing(
+            filter_, check_for_entry_type, replace_with_relationship
+        )
+
+    def _apply_only_filter(self, filter_: dict) -> dict:
+        """This method loops through the query and replaces #Only with the proper query"""
+
+        def check_for_only_filter(_, expr):
+            """Find cases where '#only is in the query."""
+            return isinstance(expr, dict) and ("#only" in expr)
+
+        def replace_only_filter(subdict: dict, prop: str, expr: dict):
+            """Replace magic key `"#only"` (added by this transformer)
+            the first part of the query first selects all the documents that have a value, in their list, that is not within the alowed values.
+            Subsequently this selection is inversed, to get the documents that only have the alowed values.
+            With this inversion also documents with "weird" values such as null or empty lists are selected.
+            The second part of the query makes sure that only documents with lists that have at least one value are selected.
+            """
+
+            if "$and" not in subdict:
+                subdict["$and"] = []
+
+            if prop == "relationships.references.data.id":
```

Comment:
Yes, I hope to make a dictionary where the keys are the prop field and a value which indicates at which nesting level the list occurs. That way, you only have to make a new entry in this dictionary to make a new field searchable with HAS ONLY. I think I will do this in a separate PR.

I was also wondering whether it would be possible to remove `_apply_relationship_filtering`  and do the renaming of the property field in the more general mapper that maps the OPTIMADE field names to the database specific field names.

Reply:
> I was also wondering whether it would be possible to remove `_apply_relationship_filtering` and do the renaming of the property field in the more general mapper that maps the OPTIMADE field names to the database specific field names.

This would be very valuable, we could definitely do this in another PR.